JEP 360/JEP 384 Sealed classes and records DDR support

theresa-m opened the pull request:
Fixes: https://github.com/eclipse/openj9/issues/9626
Related: https://github.com/eclipse/openj9/issues/9529

0.22 port: ~~https://github.com/eclipse/openj9/pull/10405~~ https://github.com/eclipse/openj9/pull/10443

This ended up being support for both sealed classes and records which was missed in Java 14.

- `cfdump -x <classfile>`
- `!dumpromclass <address>`
- `!dumpromclasslinear <address>`

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

theresa-m commented:
Manual testing:

**Test classes:**
```
import java.lang.annotation.*;

@Retention(RetentionPolicy.RUNTIME)
@Inherited()
@interface TestAnnotation {}

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE_USE)
@interface TestTypeAnnotation {}

record RecordTestAttributes(@TestAnnotation int comp1, @TestAnnotation @TestTypeAnnotation int comp2) {
  public static void main(String[] args) {
    long[] l = new long[Integer.MAX_VALUE];
  }
}

record RecordTest(int comp1, int comp2) {
  public static void main(String[] args) {
    long[] l = new long[Integer.MAX_VALUE];
  }
}

non-sealed class SubClass1 extends SealedTest {}
non-sealed class SubClass2 extends SealedTest {}

sealed class SealedTest permits SubClass1, SubClass2 {
  public static void main(String[] args) {
    long[] l = new long[Integer.MAX_VALUE];
  }
}
```

**rcdump.c changes:**

`Cfdump -x <classfile>`

Record:
```
ROM Size: 0x518 (1304)
Class Name: RecordTest
Superclass Name: java/lang/Record
Source File Name: RecordTest.java
Oracle Access Flags (0x30):
  access: default final super
J9 Access Flags (0xA800400): (final fields) (preverified) (record)
Class file version: 59.65535
Instance Shape: 0xE
Intermediate Class Data (1304 bytes): FFFFFFFFFFFFFFC0
Maximum Branch Count: 1
Interfaces (0):
Fields (2):
  Name: comp1
  Signature: I
  Access Flags (300012):
  access: private final

  Name: comp2
  Signature: I
  Access Flags (300012):
  access: private final
```

Sealed classes: ‘
```
ROM Size: 0x1D8 (472)
Class Name: SealedTest
Superclass Name: java/lang/Object
Source File Name: SealedTest.java
Oracle Access Flags (0x20):
  access: default super
J9 Access Flags (0x8800200): (preverified) (sealed)
Class file version: 59.65535
Instance Shape: 0xE
Intermediate Class Data (472 bytes): FFFFFFFFFFFFFFC0
Maximum Branch Count: 1
Interfaces (0):
Permitted subclasses (2):
  SubClass1
  SubClass2
Fields (0):
CP Shape Description:
  . I v C
  C C
```

**J9BCUtil.java changes**

`!dumpromclass <address>`

Record:
```
!dumpromclass 0x00007FD60B4B3020
ROM Size: 0x518 (1304)
Class Name: RecordTest
Superclass Name: java/lang/Record
Source File Name: RecordTest.java
Oracle Access Flags (0x30): default final super
J9 Access Flags (0xa800400): (final fields) (preverified) (record)
Class file version: 59.65535
Instance Shape: 0xe
Intermediate Class Data (1304 bytes): 7fd60b4b3020
Maximum Branch Count: 1
Interfaces (0):
Fields (2):
  Name: comp1
  Signature: I
  Access Flags (300012): private final

  Name: comp2
  Signature: I
  Access Flags (300012): private final
```

Sealed classes:
```
!dumpromclass 0x00007FD435ABB020
ROM Size: 0x1d8 (472)
Class Name: SealedTest
Superclass Name: java/lang/Object
Source File Name: SealedTest.java
Oracle Access Flags (0x20): default super
J9 Access Flags (0x8800200): (preverified) (sealed)
Class file version: 59.65535
Instance Shape: 0xe
Intermediate Class Data (472 bytes): 7fd435abb020
Maximum Branch Count: 1
Interfaces (0):
Permitted subclasses (2):
   SubClass1
   SubClass2
Fields (0):
```

**RomClassWalker.java changes**

`!dumpromclasslinear <address>`

Record:
```
!dumpromclasslinear 0x00007FD60B4B3020
ROM Class 'RecordTest' at 0x00007FD60B4B3020
0x00007FD60B4B3020-0x00007FD60B4B30D0 [            (SECTION) romHeader                    ]    176 bytes
0x00007FD60B4B30D0-0x00007FD60B4B3138 [            (SECTION) constantPool                 ]    104 bytes
0x00007FD60B4B3138-0x00007FD60B4B3150 [            (SECTION) fields                       ]     24 bytes
0x00007FD60B4B3150-0x00007FD60B4B3188 [            (SECTION) cpNamesAndSignaturesSRPs     ]     56 bytes
0x00007FD60B4B3188-0x00007FD60B4B32D4 [            (SECTION) methods                      ]    332 bytes
0x00007FD60B4B32D4-0x00007FD60B4B32DC [            (SECTION) cpShapeDescription           ]      8 bytes
0x00007FD60B4B32E4-0x00007FD60B4B32E8 [           0x00000002 numberRecordComponents       ]
0x00007FD60B4B32E8-0x00007FD60B4B32EC [           0x0000006C cpFieldNAS                   ]  -> 0x00007FD60B4B3354
0x00007FD60B4B32F4-0x00007FD60B4B32F8 [           0x0000006C cpFieldNAS                   ]  -> 0x00007FD60B4B3360
0x00007FD60B4B32DC-0x00007FD60B4B3300 [                      Padding                      ] 36 byte(s)
0x00007FD60B4B3300-0x00007FD60B4B3308 [            (SECTION) optionalInfo                 ]      8 bytes
0x00007FD60B4B3308-0x00007FD60B4B3328 [                      Padding                      ] 32 byte(s)
0x00007FD60B4B3328-0x00007FD60B4B3538 [            (SECTION) UTF8                         ]    528 bytes
```

```
!dumpromclasslinear 0x00007FB5072B3020
ROM Class 'RecordTestAttributes' at 0x00007FB5072B3020
0x00007FB5072B3020-0x00007FB5072B30D0 [            (SECTION) romHeader                    ]    176 bytes
0x00007FB5072B30D0-0x00007FB5072B3148 [            (SECTION) constantPool                 ]    120 bytes
0x00007FB5072B3148-0x00007FB5072B3178 [            (SECTION) fields                       ]     48 bytes
0x00007FB5072B3178-0x00007FB5072B3184 [                      Padding                      ] 12 byte(s)
0x00007FB5072B3184-0x00007FB5072B31BC [            (SECTION) cpNamesAndSignaturesSRPs     ]     56 bytes
0x00007FB5072B31BC-0x00007FB5072B3358 [            (SECTION) methods                      ]    412 bytes
0x00007FB5072B3358-0x00007FB5072B3360 [            (SECTION) cpShapeDescription           ]      8 bytes
0x00007FB5072B3360-0x00007FB5072B3378 [                      Padding                      ] 24 byte(s)
0x00007FB5072B3368-0x00007FB5072B336C [           0x00000002 numberRecordComponents       ]
0x00007FB5072B336C-0x00007FB5072B3370 [           0x0000009A cpFieldNAS                   ]  -> 0x00007FB5072B3406
0x00007FB5072B3378-0x00007FB5072B3384 [            (SECTION) recordComponentAnnotation    ]     12 bytes
0x00007FB5072B3384-0x00007FB5072B3388 [           0x0000008E cpFieldNAS                   ]  -> 0x00007FB5072B3412
0x00007FB5072B3384-0x00007FB5072B3390 [                      Padding                      ] 12 byte(s)
0x00007FB5072B3390-0x00007FB5072B339C [            (SECTION) recordComponentAnnotation    ]     12 bytes
0x00007FB5072B339C-0x00007FB5072B33A8 [            (SECTION) recordComponentTypeAnnotation ]     12 bytes
0x00007FB5072B33A8-0x00007FB5072B33B0 [            (SECTION) optionalInfo                 ]      8 bytes
0x00007FB5072B33B0-0x00007FB5072B33D0 [                      Padding                      ] 32 byte(s)
0x00007FB5072B33D0-0x00007FB5072B363A [            (SECTION) UTF8                         ]    618 bytes
0x00007FB5072B363A-0x00007FB5072B3640 [                      Padding                      ] 6 byte(s)
```

Sealed classes:
```
!dumpromclasslinear 0x00007FD435ABB020
ROM Class 'SealedTest' at 0x00007FD435ABB020
0x00007FD435ABB020-0x00007FD435ABB0D0 [            (SECTION) romHeader                    ]    176 bytes
0x00007FD435ABB0D0-0x00007FD435ABB100 [            (SECTION) constantPool                 ]     48 bytes
0x00007FD435ABB100-0x00007FD435ABB108 [            (SECTION) cpNamesAndSignaturesSRPs     ]      8 bytes
0x00007FD435ABB108-0x00007FD435ABB160 [            (SECTION) methods                      ]     88 bytes
0x00007FD435ABB160-0x00007FD435ABB164 [            (SECTION) cpShapeDescription           ]      4 bytes
0x00007FD435ABB164-0x00007FD435ABB174 [                      Padding                      ] 16 byte(s)
0x00007FD435ABB168-0x00007FD435ABB16C [           0x00000002 numberPermittedSubclasses    ]
0x00007FD435ABB16C-0x00007FD435ABB170 [           0x00000070 permittedSubclassName        ]  -> 0x00007FD435ABB1DC SubClass1
0x00007FD435ABB170-0x00007FD435ABB174 [           0x00000078 permittedSubclassName        ]  -> 0x00007FD435ABB1E8 SubClass2
0x00007FD435ABB174-0x00007FD435ABB17C [            (SECTION) optionalInfo                 ]      8 bytes
0x00007FD435ABB17C-0x00007FD435ABB180 [                      Padding                      ] 4 byte(s)
0x00007FD435ABB180-0x00007FD435ABB1F4 [            (SECTION) UTF8                         ]    116 bytes
0x00007FD435ABB1F4-0x00007FD435ABB1F8 [                      Padding                      ] 4 byte(s)
```

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -30,6 +30,7 @@
 import static com.ibm.j9ddr.vm29.structure.J9NonbuilderConstants.J9_ROMCLASS_OPTINFO_SOURCE_DEBUG_EXTENSION;
 import static com.ibm.j9ddr.vm29.structure.J9NonbuilderConstants.J9_ROMCLASS_OPTINFO_SOURCE_FILE_NAME;
 import static com.ibm.j9ddr.vm29.structure.J9NonbuilderConstants.J9_ROMCLASS_OPTINFO_TYPE_ANNOTATION_INFO;
+import static com.ibm.j9ddr.vm29.structure.J9NonbuilderConstants.J9_ROMCLASS_OPTINFO_PERMITTEDSUBCLASSES_ATTRIBUTE;
```
Please keep imports sorted.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -236,4 +237,33 @@ public static J9SourceDebugExtensionPointer getSourceDebugExtensionForROMClass(J
 	}
 
+	private static U32Pointer getPermittedSubclassPointer(J9ROMClassPointer romClass) throws CorruptDataException {
+		SelfRelativePointer srpPtr = getSRPPtr(J9ROMClassHelper.optionalInfo(romClass), romClass.optionalFlags(), 
+			J9_ROMCLASS_OPTINFO_PERMITTEDSUBCLASSES_ATTRIBUTE);
+		if (!srpPtr.isNull()) {
+			return U32Pointer.cast(srpPtr.get());
+		}	
```
Please remove trailing whitespace (at least in new code).

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -846,6 +860,54 @@ int allSlotsInVerificationTypeInfoDo(U8Pointer cursor) throws CorruptDataExcepti
 		}
 		return 3;
 	}
+	void recordAttributeDo(U32Pointer attribute) throws CorruptDataException
```
Please leave a blank line between methods.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/J9BCUtil.java:
```diff
@@ -557,10 +557,10 @@ public static void j9bcutil_dumpRomClass(PrintStream out, J9ROMClassPointer romC
 		dumpEnclosingMethod(out, romClass, flags);
 
-		out.append(String.format("Sun Access Flags (0x%s): ", Long.toHexString(romClass.modifiers().longValue())));
+		out.append(String.format("Oracle Access Flags (0x%s): ", Long.toHexString(romClass.modifiers().longValue())));
```
Have you verified that no tests are affected by this change?

theresa-m replied:
No. I will try it out.

theresa-m replied:
The OpenJ9 DDR tests are passing for me locally.

keithc-ca replied:
As I think about this again, I think (given we're changing things) that it should not refer to 'Sun' or 'Oracle': How about 'Basic Access Flags'?

theresa-m replied:
Agreed. I originally added Oracle to match the statement in J9BCUtil.java but I think both should be updated.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/J9BCUtil.java:
```diff
@@ -610,6 +610,18 @@ public static void j9bcutil_dumpRomClass(PrintStream out, J9ROMClassPointer romC
+		/* Permitted subclasses for a sealed class */
+		if (J9ROMClassHelper.isSealed(romClass)) {
+			int permittedSubclassCount = OptInfo.getPermittedSubclassCount(romClass);
+			out.append(String.format("Permitted subclasses (%d):" + nl, permittedSubclassCount));
```
Please instead use `out.format("Permitted subclasses (%d):%n", permittedSubclassCount)`
* `%n` expresses the desired newline
* there's no need to build the string locally

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/J9BCUtil.java:
```diff
@@ -610,6 +610,18 @@ public static void j9bcutil_dumpRomClass(PrintStream out, J9ROMClassPointer romC
+
+			for (int i = 0; i < permittedSubclassCount; i++) {
+				J9UTF8Pointer permittedSubclassName = OptInfo.getPermittedSubclassNameAtIndex(romClass, i);
+				out.append("   " + J9UTF8Helper.stringValue(permittedSubclassName));
+				out.append(nl);
```
`out.format("   %s%n", J9UTF8Helper.stringValue(permittedSubclassName));`

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/J9BCUtil.java:
```diff
@@ -715,6 +727,10 @@ private static void dumpClassJ9ExtraModifiers(PrintStream out, long accessFlags)
 		if ((accessFlags & J9AccClassIsUnmodifiableBit) != 0)
 			out.append("(unmodifiable) ");
+		if ((accessFlags & J9JavaAccessFlags.J9AccRecord) != 0)
+			out.append("(record)");
+		if ((accessFlags & J9JavaAccessFlags.J9AccSealed) != 0)
+			out.append("(sealed)");
```
Spaces after closing `)` are missing.

keithc-ca commented on runtime/util/rcdump.c:
```diff
@@ -180,6 +181,11 @@ IDATA j9bcutil_dumpRomClass( J9ROMClass *romClass, J9PortLibrary *portLib, J9Tra
 	}
 
+	if (J9ROMCLASS_IS_SEALED(romClass))
+	{
```
formatting: please move brace to end of preceding line.

keithc-ca commented on runtime/util/rcdump.c:
```diff
@@ -823,6 +829,22 @@ dumpEnclosingMethod(J9PortLibrary *portLib, J9ROMClass *romClass, U_32 flags)
+
+	j9tty_printf(PORTLIB, "Permitted subclasses (%i):\n", *permittedSubclassesCountPtr);
+	for (; i < *permittedSubclassesCountPtr; i++) {
+		J9UTF8* permittedSubclassNameUtf8 = permittedSubclassesNameAtIndex(permittedSubclassesCountPtr, i);
```
Please be consistent: `*` next to name instead of type:
```
J9UTF8 *permittedSubclassNameUtf8 = ...
```

keithc-ca commented on runtime/util/rcdump.c:
```diff
@@ -823,6 +829,22 @@ dumpEnclosingMethod(J9PortLibrary *portLib, J9ROMClass *romClass, U_32 flags)
+{
+	PORT_ACCESS_FROM_PORT(portLib);
+
+	U_32 *permittedSubclassesCountPtr = getNumberOfPermittedSubclassesPtr(romClass);
```
`getNumberOfPermittedSubclassesPtr` seems a strange API to me: why return a pointer (nobody appears to write to it)?
If we decide to change it, that change should be deferred to a separate pull request.

theresa-m replied:
Yes its a bit odd to the other optinfo api methods. Other variable-length sections of ROM class optionalInfo have iterator like setups but that seemed like overkill to me here since its only a list of class names that is stored.

The list of permitted subclass names is located directly after `permittedSubclassesCountPtr` so `permittedSubclassesNameAtIndex` takes in this pointer later as a starting address to find the class name.

keithc-ca commented on runtime/util/rcdump.c:
```diff
@@ -1325,6 +1347,20 @@ j9_printClassExtraModifiers(J9PortLibrary *portLib, U_32 modifiers)
 		if(modifiers) j9tty_printf(PORTLIB, " ");
 	}
+
+	if(modifiers & J9AccRecord)
+	{
```
formatting (here and line 1358)

keithc-ca commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -108,6 +108,8 @@
 #define J9FieldTypeMask 0x380000
 #define J9FieldTypeShort 0x280000
 
+/* @ddr_namespace: map_to_type=J9RecordComponentFlags */
```
Moving the constants below to a new namespace creates a discontinuity in core files.
The change seems reasonable, but I please confirm you did so consciously.

theresa-m replied:
I am not familiar with the consequences of this change. What do you recommend to confirm its okay?

keithc-ca replied:
A core file generated from a VM built before this change would have these constants in `J9FieldFlags`.
I think `J9RecordComponentFlags` is a better place for them, but it will mean `NoSuchFieldError` will be thrown by DDR code that makes use of `J9RecordComponentFlags.J9RecordComponentFlagHasAnnotations` (for example) when accessing such an old core file. Because records are still a preview feature, I can live with that: My point is that it needs to be a conscious decision.

theresa-m replied:
Okay. Thanks for the clarification.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -236,4 +237,33 @@ public static J9SourceDebugExtensionPointer getSourceDebugExtensionForROMClass(J
 	}
 
+	private static U32Pointer getPermittedSubclassPointer(J9ROMClassPointer romClass) throws CorruptDataException {
+		SelfRelativePointer srpPtr = getSRPPtr(J9ROMClassHelper.optionalInfo(romClass), romClass.optionalFlags(), 
```
Please remove all trailing whitespace in code you're adding (e.g. this line).

keithc-ca commented:
In the examples given, the padding doesn't seem right to me:
```
!dumpromclasslinear 0x00007FD60B4B3020
ROM Class 'RecordTest' at 0x00007FD60B4B3020
0x00007FD60B4B3020-0x00007FD60B4B30D0 [            (SECTION) romHeader                    ]    176 bytes
0x00007FD60B4B30D0-0x00007FD60B4B3138 [            (SECTION) constantPool                 ]    104 bytes
0x00007FD60B4B3138-0x00007FD60B4B3150 [            (SECTION) fields                       ]     24 bytes
0x00007FD60B4B3150-0x00007FD60B4B3188 [            (SECTION) cpNamesAndSignaturesSRPs     ]     56 bytes
0x00007FD60B4B3188-0x00007FD60B4B32D4 [            (SECTION) methods                      ]    332 bytes
0x00007FD60B4B32D4-0x00007FD60B4B32DC [            (SECTION) cpShapeDescription           ]      8 bytes
0x00007FD60B4B32E4-0x00007FD60B4B32E8 [           0x00000002 numberRecordComponents       ]
0x00007FD60B4B32E8-0x00007FD60B4B32EC [           0x0000006C cpFieldNAS                   ]  -> 0x00007FD60B4B3354
0x00007FD60B4B32F4-0x00007FD60B4B32F8 [           0x0000006C cpFieldNAS                   ]  -> 0x00007FD60B4B3360
0x00007FD60B4B32DC-0x00007FD60B4B3300 [                      Padding                      ] 36 byte(s)
0x00007FD60B4B3300-0x00007FD60B4B3308 [            (SECTION) optionalInfo                 ]      8 bytes
0x00007FD60B4B3308-0x00007FD60B4B3328 [                      Padding                      ] 32 byte(s)
0x00007FD60B4B3328-0x00007FD60B4B3538 [            (SECTION) UTF8                         ]    528 bytes
```
I think this padding is missing:
```
0x00007FD60B4B32DC-0x00007FD60B4B32E4 [                      Padding                      ] 8 byte(s)
```
and the first padding line above should be:
```
0x00007FD60B4B32F8-0x00007FD60B4B3300 [                      Padding                      ] 8 byte(s)
```
Isn't there real information present instead of this padding?
```
0x00007FD60B4B3308-0x00007FD60B4B3328 [                      Padding                      ] 32 byte(s)
```

theresa-m commented:
Ah yes I am missing the attributeFlags from J9ROMRecordComponentShape. I also added some logic to calculate extra padding needed for the record component.

It looks like the other sections of optionalInfo are all in their own sections so I added the record and permittedsubclass attributes as sections as well.

This is what the output looks like now:
```
> !dumpromclasslinear 0x00007FD60B4B3020
ROM Class 'RecordTest' at 0x00007FD60B4B3020
0x00007FD60B4B3020-0x00007FD60B4B30D0 [            (SECTION) romHeader                    ]    176 bytes
0x00007FD60B4B30D0-0x00007FD60B4B3138 [            (SECTION) constantPool                 ]    104 bytes
0x00007FD60B4B3138-0x00007FD60B4B3150 [            (SECTION) fields                       ]     24 bytes
0x00007FD60B4B3150-0x00007FD60B4B3188 [            (SECTION) cpNamesAndSignaturesSRPs     ]     56 bytes
0x00007FD60B4B3188-0x00007FD60B4B32D4 [            (SECTION) methods                      ]    332 bytes
0x00007FD60B4B32D4-0x00007FD60B4B32DC [            (SECTION) cpShapeDescription           ]      8 bytes
0x00007FD60B4B32DC-0x00007FD60B4B32E4 [                      Padding                      ] 8 byte(s)
0x00007FD60B4B32E4-0x00007FD60B4B3300 [            (SECTION) recordComponent              ]     28 bytes
0x00007FD60B4B3300-0x00007FD60B4B3308 [            (SECTION) optionalInfo                 ]      8 bytes
0x00007FD60B4B3308-0x00007FD60B4B3328 [                      Padding                      ] 32 byte(s)
0x00007FD60B4B3328-0x00007FD60B4B3538 [            (SECTION) UTF8                         ]    528 bytes
```

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -846,6 +860,76 @@ int allSlotsInVerificationTypeInfoDo(U8Pointer cursor) throws CorruptDataExcepti
+			J9ROMRecordComponentShapePointer recordComponent = J9ROMRecordComponentShapePointer.cast(attribute);
+			attribute = U32Pointer.cast(recordComponent.add(1));
+
+			/* name and descriptor */
```
This comment is stale, and I'm not sure it added much value.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -846,6 +860,76 @@ int allSlotsInVerificationTypeInfoDo(U8Pointer cursor) throws CorruptDataExcepti
+				attribute = attribute.add(increment);
+			}
+		}
+		
```
Please remove the trailing whitespace.

keithc-ca replied:
It's still here.

theresa-m replied:
Thanks. It is removed now.

keithc-ca replied:
I see you've completely removed the lines containing only whitespace: it's not what I was asking for, but that's ok.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -236,4 +237,33 @@ public static J9SourceDebugExtensionPointer getSourceDebugExtensionForROMClass(J
+	private static U32Pointer getPermittedSubclassPointer(J9ROMClassPointer romClass) throws CorruptDataException {
+		SelfRelativePointer srpPtr = getSRPPtr(J9ROMClassHelper.optionalInfo(romClass), romClass.optionalFlags(),
+			J9_ROMCLASS_OPTINFO_PERMITTEDSUBCLASSES_ATTRIBUTE);
+		if (!srpPtr.isNull()) {
```
Suggest you say `srpPtr.notNull()`, or invert.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -236,4 +237,33 @@ public static J9SourceDebugExtensionPointer getSourceDebugExtensionForROMClass(J
+
+	public static int getPermittedSubclassCount(J9ROMClassPointer romClass) throws CorruptDataException {
+		U32Pointer permittedSubclassPointer = getPermittedSubclassPointer(romClass);
+		if (!permittedSubclassPointer.isNull()) {
```
Suggest `permittedSubclassPointer.notNull()` or invert.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/j9/OptInfo.java:
```diff
@@ -236,4 +237,33 @@ public static J9SourceDebugExtensionPointer getSourceDebugExtensionForROMClass(J
+
+	public static J9UTF8Pointer getPermittedSubclassNameAtIndex(J9ROMClassPointer romClass, int index) throws CorruptDataException {
+		U32Pointer permittedSubclassPointer = getPermittedSubclassPointer(romClass);
+		if (!permittedSubclassPointer.isNull()) {
```
Suggest `permittedSubclassPointer.notNull()` or invert.

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -687,6 +691,16 @@ void allSlotsInOptionalInfoDo() throws CorruptDataException
 			cursor = cursor.add(1);
 		}
+		if (romClass.optionalFlags().allBitsIn(J9NonbuilderConstants.J9_ROMCLASS_OPTINFO_RECORD_ATTRIBUTE)) {
+			classWalkerCallback.addSlot(clazz, SlotType.J9_SRP, cursor, "RecordAttributeSRP");
```
Most calls to `addSlot()` supply a slot name beginning with a lower-case letter: why the departure here and on line 700 below?

theresa-m replied:
I will change it to lower case.

theresa-m commented:
I have created a PR from this branch targeted to 0.22: https://github.com/eclipse/openj9/pull/10405

keithc-ca commented:
`!dumpromclasslinear` doesn't work as expected: I don't see `numberRecordComponents` or `cpFieldNAS`:
```
> !dumpromclasslinear 0x00007F652033C030
ROM Class 'RecordTest' at 0x00007F652033C030
0x00007F652033C030-0x00007F652033C0E0 [            (SECTION) romHeader                    ]    176 bytes
0x00007F652033C0E0-0x00007F652033C178 [            (SECTION) constantPool                 ]    152 bytes
0x00007F652033C178-0x00007F652033C190 [            (SECTION) fields                       ]     24 bytes
0x00007F652033C190-0x00007F652033C1D0 [            (SECTION) cpNamesAndSignaturesSRPs     ]     64 bytes
0x00007F652033C1D0-0x00007F652033C328 [            (SECTION) methods                      ]    344 bytes
0x00007F652033C328-0x00007F652033C334 [            (SECTION) cpShapeDescription           ]     12 bytes
0x00007F652033C334-0x00007F652033C33C [                      Padding                      ] 8 byte(s)
0x00007F652033C33C-0x00007F652033C358 [            (SECTION) recordComponent              ]     28 bytes
0x00007F652033C358-0x00007F652033C360 [            (SECTION) optionalInfo                 ]      8 bytes
0x00007F652033C360-0x00007F652033C380 [                      Padding                      ] 32 byte(s)
0x00007F652033C380-0x00007F652033C536 [            (SECTION) UTF8                         ]    438 bytes
0x00007F652033C536-0x00007F652033C538 [                      Padding                      ] 2 byte(s)
```

theresa-m commented:
> `!dumpromclasslinear` doesn't work as expected: I don't see `numberRecordComponents` or `cpFieldNAS`:
> 
> ```
> > !dumpromclasslinear 0x00007F652033C030
> ROM Class 'RecordTest' at 0x00007F652033C030
> 0x00007F652033C030-0x00007F652033C0E0 [            (SECTION) romHeader                    ]    176 bytes
> 0x00007F652033C0E0-0x00007F652033C178 [            (SECTION) constantPool                 ]    152 bytes
> 0x00007F652033C178-0x00007F652033C190 [            (SECTION) fields                       ]     24 bytes
> 0x00007F652033C190-0x00007F652033C1D0 [            (SECTION) cpNamesAndSignaturesSRPs     ]     64 bytes
> 0x00007F652033C1D0-0x00007F652033C328 [            (SECTION) methods                      ]    344 bytes
> 0x00007F652033C328-0x00007F652033C334 [            (SECTION) cpShapeDescription           ]     12 bytes
> 0x00007F652033C334-0x00007F652033C33C [                      Padding                      ] 8 byte(s)
> 0x00007F652033C33C-0x00007F652033C358 [            (SECTION) recordComponent              ]     28 bytes
> 0x00007F652033C358-0x00007F652033C360 [            (SECTION) optionalInfo                 ]      8 bytes
> 0x00007F652033C360-0x00007F652033C380 [                      Padding                      ] 32 byte(s)
> 0x00007F652033C380-0x00007F652033C536 [            (SECTION) UTF8                         ]    438 bytes
> 0x00007F652033C536-0x00007F652033C538 [                      Padding                      ] 2 byte(s)
> ```

I added a section around the entire record component since that seems to be the standard way to display and checked that those lines were showing up with the section part commented out. I've not been able to figure out how to display information from the individual sections though.. do you know the command for that?

keithc-ca commented:
`!dumpromclasslinear <address>,N` displays extra levels of detail when `N` is greater than 1:
```
=== Section Start: recordComponent (28 bytes)                                             ===
0x00007F93005A94EC-0x00007F93005A94F0 [           0x00000002 numberRecordComponents       ] 

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007F93005A94F0-0x00007F93005A94F4 [           0x0000005E cpFieldNAS                   ]  -> 0x00007F93005A954E
0x00007F93005A94F4-0x00007F93005A94F8 [                      Padding                      ] 4 byte(s)
0x00007F93005A94F8-0x00007F93005A94FC [           0x00000000 attributeFlags               ] 

=== Section End: recordComponentShape                                                     ===

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007F93005A94FC-0x00007F93005A9500 [           0x0000005A cpFieldNAS                   ]  -> 0x00007F93005A9556
0x00007F93005A9500-0x00007F93005A9504 [                      Padding                      ] 4 byte(s)
0x00007F93005A9504-0x00007F93005A9508 [           0x00000000 attributeFlags               ] 

=== Section End: recordComponentShape                                                     ===

=== Section End: recordComponent                                                          ===
```
It would be helpful to provide more detail for `cpFieldNAS`.

I also see some troubling output - memory faults and negative padding:
```
=== Section Start: UTF8 (438 bytes)                                                       ===
0x00007F93005A9530-0x00007F93005A9542 [                 UTF8 cpFieldUtf8                  ] java/lang/Record
0x00007F93005A9542-0x00007F93005A954E [                 UTF8 cpFieldUtf8                  ] RecordTest
Memory Fault reading 0x7F936FBD9553 
0x00007F93005A9552-0x00007F93005A954E [                      Padding                      ] -4 byte(s)
0x00007F93005A954E-0x00007F93005A9556 [                 UTF8 name                         ] comp1
0x00007F93005A9556-0x00007F93005A9552 [                      Padding                      ] -4 byte(s)
Memory Fault reading 0x7F93008C05BF : MemoryFault loading cache block, unbacked memory
Memory Fault reading 0x7F936FBD955B 
0x00007F93005A955A-0x00007F93005A9556 [                      Padding                      ] -4 byte(s)
0x00007F93005A9556-0x00007F93005A955E [                 UTF8 name                         ] comp2
0x00007F93005A955E-0x00007F93005A955A [                      Padding                      ] -4 byte(s)
Memory Fault reading 0x7F93008D05C7 : MemoryFault loading cache block, unbacked memory
0x00007F93005A955E-0x00007F93005A9574 [                 UTF8 cpFieldUtf8                  ] RecordTestAttributes
0x00007F93005A9574-0x00007F93005A9580 [                 UTF8 cpFieldUtf8                  ] SubClass1
0x00007F93005A9580-0x00007F93005A958C [                 UTF8 cpFieldUtf8                  ] SubClass2
0x00007F93005A958C-0x00007F93005A95E4 [                      Padding                      ] 88 byte(s)
0x00007F93005A95E4-0x00007F93005A95F6 [                 UTF8 optionalFileNameUTF8         ] RecordTest.java
0x00007F93005A95F6-0x00007F93005A9618 [                 UTF8 cpFieldUtf8                  ] java/lang/runtime/ObjectMethods
0x00007F93005A9618-0x00007F93005A9624 [                 UTF8 name                         ] bootstrap
0x00007F93005A9624-0x00007F93005A96D8 [                 UTF8 signature                    ] (Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/TypeDescriptor;Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/invoke/MethodHandle;)Ljava/lang/Object;
0x00007F93005A96D8-0x00007F93005A96E6 [                 UTF8 cpFieldUtf8                  ] comp1;comp2

=== Section End: UTF8                                                                     ===
```

theresa-m commented:
I starting printing the NAS components in J9ROMRecordComponentShapePointer separately (see latest commit) and things are looking better:

Results for record test without optional attributes:

```
=== Section Start: recordComponent (28 bytes)                                             ===
0x00007F9667DBB2E4-0x00007F9667DBB2E8 [           0x00000002 numberRecordComponents       ]

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007F9667DBB2E8-0x00007F9667DBB2EC [           0x0000006C name                         ]  -> 0x00007F9667DBB354 comp1
0x00007F9667DBB2EC-0x00007F9667DBB2F0 [           0x00000070 signature                    ]  -> 0x00007F9667DBB35C I
0x00007F9667DBB2F0-0x00007F9667DBB2F4 [           0x00000000 attributeFlags               ]

=== Section End: recordComponentShape                                                     ===

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007F9667DBB2F4-0x00007F9667DBB2F8 [           0x0000006C name                         ]  -> 0x00007F9667DBB360 comp2
0x00007F9667DBB2F8-0x00007F9667DBB2FC [           0x00000064 signature                    ]  -> 0x00007F9667DBB35C I
0x00007F9667DBB2FC-0x00007F9667DBB300 [           0x00000000 attributeFlags               ]

=== Section End: recordComponentShape                                                     ===

=== Section End: recordComponent

…

=== Section Start: UTF8 (528 bytes)                                                       ===
0x00007F9667DBB328-0x00007F9667DBB33A [                 UTF8 cpFieldUtf8                  ] java/lang/Record
0x00007F9667DBB33A-0x00007F9667DBB342 [                 UTF8 name                         ] <init>
0x00007F9667DBB342-0x00007F9667DBB348 [                 UTF8 signature                    ] ()V
0x00007F9667DBB348-0x00007F9667DBB354 [                 UTF8 cpFieldUtf8                  ] RecordTest
0x00007F9667DBB354-0x00007F9667DBB35C [                 UTF8 name                         ] comp1
0x00007F9667DBB35C-0x00007F9667DBB360 [                 UTF8 signature                    ] I
0x00007F9667DBB360-0x00007F9667DBB368 [                 UTF8 name                         ] comp2
0x00007F9667DBB368-0x00007F9667DBB372 [                 UTF8 name                         ] toString
0x00007F9667DBB372-0x00007F9667DBB394 [                      Padding                      ] 34 byte(s)
0x00007F9667DBB394-0x00007F9667DBB39E [                 UTF8 name                         ] hashCode
0x00007F9667DBB39E-0x00007F9667DBB3B0 [                      Padding                      ] 18 byte(s)
0x00007F9667DBB3B0-0x00007F9667DBB3B8 [                 UTF8 name                         ] equals
0x00007F9667DBB3B8-0x00007F9667DBB3DC [                      Padding                      ] 36 byte(s)
0x00007F9667DBB3DC-0x00007F9667DBB3E4 [                 UTF8 signature                    ] (II)V
0x00007F9667DBB3E4-0x00007F9667DBB3EA [                 UTF8 name                         ] main
0x00007F9667DBB3EA-0x00007F9667DBB402 [                 UTF8 signature                    ] ([Ljava/lang/String;)V
0x00007F9667DBB402-0x00007F9667DBB418 [                 UTF8 signature                    ] ()Ljava/lang/String;
0x00007F9667DBB418-0x00007F9667DBB41E [                 UTF8 signature                    ] ()I
0x00007F9667DBB41E-0x00007F9667DBB436 [                 UTF8 signature                    ] (Ljava/lang/Object;)Z
0x00007F9667DBB436-0x00007F9667DBB448 [                 UTF8 optionalFileNameUTF8         ] RecordTest.java
0x00007F9667DBB448-0x00007F9667DBB46A [                 UTF8 cpFieldUtf8                  ] java/lang/runtime/ObjectMethods
0x00007F9667DBB46A-0x00007F9667DBB476 [                 UTF8 name                         ] bootstrap
0x00007F9667DBB476-0x00007F9667DBB52A [                 UTF8 signature                    ] (Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/TypeDescriptor;Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/invoke/MethodHandle;)Ljava/lang/Object;
0x00007F9667DBB52A-0x00007F9667DBB538 [                 UTF8 cpFieldUtf8                  ] comp1;comp2

=== Section End: UTF8
```

Results for record test with attribtues:
```
=== Section Start: recordComponent (64 bytes)                                             ===
0x00007FC8A3CA4368-0x00007FC8A3CA436C [           0x00000002 numberRecordComponents       ]

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007FC8A3CA436C-0x00007FC8A3CA4370 [           0x0000009A name                         ]  -> 0x00007FC8A3CA4406 comp1
0x00007FC8A3CA4370-0x00007FC8A3CA4374 [           0x0000009E signature                    ]  -> 0x00007FC8A3CA440E I
0x00007FC8A3CA4374-0x00007FC8A3CA4378 [           0x00000002 attributeFlags               ]

=== Section End: recordComponentShape                                                     ===

=== Section Start: recordComponentAnnotation (12 bytes)                                   ===
0x00007FC8A3CA4378-0x00007FC8A3CA437C [           0x00000006 annotation length            ]
0x00007FC8A3CA437C-0x00007FC8A3CA437D [                 0x00 annotation data              ]
0x00007FC8A3CA437D-0x00007FC8A3CA437E [                 0x01 annotation data              ]
0x00007FC8A3CA437E-0x00007FC8A3CA437F [                 0x00 annotation data              ]
0x00007FC8A3CA437F-0x00007FC8A3CA4380 [                 0x07 annotation data              ]
0x00007FC8A3CA4380-0x00007FC8A3CA4381 [                 0x00 annotation data              ]
0x00007FC8A3CA4381-0x00007FC8A3CA4382 [                 0x00 annotation data              ]
0x00007FC8A3CA4382-0x00007FC8A3CA4383 [                 0x00 annotation padding           ]
0x00007FC8A3CA4383-0x00007FC8A3CA4384 [                 0x00 annotation padding           ]

=== Section End: recordComponentAnnotation                                                ===

=== Section Start: recordComponentShape (12 bytes)                                        ===
0x00007FC8A3CA4384-0x00007FC8A3CA4388 [           0x0000008E name                         ]  -> 0x00007FC8A3CA4412 comp2
0x00007FC8A3CA4388-0x00007FC8A3CA438C [           0x00000086 signature                    ]  -> 0x00007FC8A3CA440E I
0x00007FC8A3CA438C-0x00007FC8A3CA4390 [           0x00000006 attributeFlags               ]

=== Section End: recordComponentShape                                                     ===

=== Section Start: recordComponentAnnotation (12 bytes)                                   ===
0x00007FC8A3CA4390-0x00007FC8A3CA4394 [           0x00000006 annotation length            ]
0x00007FC8A3CA4394-0x00007FC8A3CA4395 [                 0x00 annotation data              ]
0x00007FC8A3CA4395-0x00007FC8A3CA4396 [                 0x01 annotation data              ]
0x00007FC8A3CA4396-0x00007FC8A3CA4397 [                 0x00 annotation data              ]
0x00007FC8A3CA4397-0x00007FC8A3CA4398 [                 0x07 annotation data              ]
0x00007FC8A3CA4398-0x00007FC8A3CA4399 [                 0x00 annotation data              ]
0x00007FC8A3CA4399-0x00007FC8A3CA439A [                 0x00 annotation data              ]
0x00007FC8A3CA439A-0x00007FC8A3CA439B [                 0x00 annotation padding           ]
0x00007FC8A3CA439B-0x00007FC8A3CA439C [                 0x00 annotation padding           ]

=== Section End: recordComponentAnnotation                                                ===

=== Section Start: recordComponentTypeAnnotation (12 bytes)                               ===
0x00007FC8A3CA439C-0x00007FC8A3CA43A0 [           0x00000008 annotation length            ]
0x00007FC8A3CA43A0-0x00007FC8A3CA43A1 [                 0x00 annotation data              ]
0x00007FC8A3CA43A1-0x00007FC8A3CA43A2 [                 0x01 annotation data              ]
0x00007FC8A3CA43A2-0x00007FC8A3CA43A3 [                 0x13 annotation data              ]
0x00007FC8A3CA43A3-0x00007FC8A3CA43A4 [                 0x00 annotation data              ]
0x00007FC8A3CA43A4-0x00007FC8A3CA43A5 [                 0x00 annotation data              ]
0x00007FC8A3CA43A5-0x00007FC8A3CA43A6 [                 0x08 annotation data              ]
0x00007FC8A3CA43A6-0x00007FC8A3CA43A7 [                 0x00 annotation data              ]
0x00007FC8A3CA43A7-0x00007FC8A3CA43A8 [                 0x00 annotation data              ]

=== Section End: recordComponentTypeAnnotation                                            ===

=== Section End: recordComponent

…

=== Section Start: UTF8 (618 bytes)                                                       ===
0x00007FC8A3CA43D0-0x00007FC8A3CA43E2 [                 UTF8 cpFieldUtf8                  ] java/lang/Record
0x00007FC8A3CA43E2-0x00007FC8A3CA43EA [                 UTF8 name                         ] <init>
0x00007FC8A3CA43EA-0x00007FC8A3CA43F0 [                 UTF8 signature                    ] ()V
0x00007FC8A3CA43F0-0x00007FC8A3CA4406 [                 UTF8 cpFieldUtf8                  ] RecordTestAttributes
0x00007FC8A3CA4406-0x00007FC8A3CA440E [                 UTF8 name                         ] comp1
0x00007FC8A3CA440E-0x00007FC8A3CA4412 [                 UTF8 signature                    ] I
0x00007FC8A3CA4412-0x00007FC8A3CA441A [                 UTF8 name                         ] comp2
0x00007FC8A3CA441A-0x00007FC8A3CA4424 [                 UTF8 name                         ] toString
0x00007FC8A3CA4424-0x00007FC8A3CA4450 [                      Padding                      ] 44 byte(s)
0x00007FC8A3CA4450-0x00007FC8A3CA445A [                 UTF8 name                         ] hashCode
0x00007FC8A3CA445A-0x00007FC8A3CA4476 [                      Padding                      ] 28 byte(s)
0x00007FC8A3CA4476-0x00007FC8A3CA447E [                 UTF8 name                         ] equals
0x00007FC8A3CA447E-0x00007FC8A3CA44AC [                      Padding                      ] 46 byte(s)
0x00007FC8A3CA44AC-0x00007FC8A3CA44BE [                 UTF8 cpFieldUtf8                  ] LTestAnnotation;
0x00007FC8A3CA44BE-0x00007FC8A3CA44D4 [                 UTF8 cpFieldUtf8                  ] LTestTypeAnnotation;
0x00007FC8A3CA44D4-0x00007FC8A3CA44DC [                 UTF8 signature                    ] (II)V
0x00007FC8A3CA44DC-0x00007FC8A3CA44E2 [                 UTF8 name                         ] main
0x00007FC8A3CA44E2-0x00007FC8A3CA44FA [                 UTF8 signature                    ] ([Ljava/lang/String;)V
0x00007FC8A3CA44FA-0x00007FC8A3CA4510 [                 UTF8 signature                    ] ()Ljava/lang/String;
0x00007FC8A3CA4510-0x00007FC8A3CA4516 [                 UTF8 signature                    ] ()I
0x00007FC8A3CA4516-0x00007FC8A3CA452E [                 UTF8 signature                    ] (Ljava/lang/Object;)Z
0x00007FC8A3CA452E-0x00007FC8A3CA454A [                 UTF8 optionalFileNameUTF8         ] RecordTestAttributes.java
0x00007FC8A3CA454A-0x00007FC8A3CA456C [                 UTF8 cpFieldUtf8                  ] java/lang/runtime/ObjectMethods
0x00007FC8A3CA456C-0x00007FC8A3CA4578 [                 UTF8 name                         ] bootstrap
0x00007FC8A3CA4578-0x00007FC8A3CA462C [                 UTF8 signature                    ] (Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/TypeDescriptor;Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/invoke/MethodHandle;)Ljava/lang/Object;
0x00007FC8A3CA462C-0x00007FC8A3CA463A [                 UTF8 cpFieldUtf8                  ] comp1;comp2

=== Section End: UTF8
```

keithc-ca commented on debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/RomClassWalker.java:
```diff
@@ -846,6 +860,72 @@ int allSlotsInVerificationTypeInfoDo(U8Pointer cursor) throws CorruptDataExcepti
+			classWalkerCallback.addSlot(clazz, SlotType.J9_ROM_UTF8, recordComponentNAS.nameEA(), "name");
+			classWalkerCallback.addSlot(clazz, SlotType.J9_ROM_UTF8, recordComponentNAS.signatureEA(), "signature");
+			classWalkerCallback.addSlot(clazz, SlotType.J9_U32, recordComponent.attributeFlagsEA(), "attributeFlags");
+			classWalkerCallback.addSection(clazz, recordComponent, attribute.getAddress() - recordComponent.getAddress(), "recordComponentShape", true);
```
The length could be expressed more clearly as `J9ROMRecordComponentShape.SIZEOF`.

theresa-m replied:
updated with this change.

keithc-ca commented:
Please squash.

theresa-m commented:
Squashed.

keithc-ca commented:
Jenkins test sanity,extended aix,osx,plinux,win jdk11

theresa-m commented:
New 0.22 port PR: https://github.com/eclipse/openj9/pull/10443
